llvm/test/CodeGen/PowerPC/aix-vec_insert_elt.ll

```diff
@@ -750,25 +750,21 @@ entry:
 define <2 x double> @testDoubleImm1(<2 x double> %a, double %b) {
 ; CHECK-64-LABEL: testDoubleImm1:
 ; CHECK-64:       # %bb.0: # %entry
-; CHECK-64-NEXT:    # kill: def $f1 killed $f1 def $vsl1
```

Comment:
this looks like degradation.
```
> renamable $f1 = COPY $f1, implicit-def $vsl1
Identity copy: renamable $f1 = COPY $f1, implicit-def $vsl1
  replace by: renamable $f1 = KILL $f1, implicit-def $vsl1
```
The `Kill` pseudo opcode is inserted from this register rewrite.

`Kill` can tell the later pass that the f1/vsl1 is killed before instruction `COPY`. See `VirtRegRewriter::handleIdentityCopy()`.  I don't think adding phony registers for the high part of vsl1 should break this part.

Reply:
> Kill can tell the later pass that the f1/vsl1 is killed before instruction COPY

I think f1/vsl1 is killed at the COPY, not before, so that this identity COPY is replaced by a `KILL` instruction. I'm wondering how we get rid of the identity COPY after adding phony registers.

Reply:
I took a look at this and the copy has changed which is why the kill is no longer emitted. 
Before the change:
```
renamable $f1 = COPY $f1, implicit-def $vsl1
```
After the change:
```
renamable $f1 = COPY $f1
```
Basically, we are missing the `implicit-def`. There is nothing wrong with not having the kill if the `implicit-def` is not there. However, I'm not sure if the `implicit-def` is necessary here or not.  

A few more details here. This is as intended. As you mentioned, the kill is added by `handleIdentityCopy()` and is guarded by `if (MI.getOperand(1).isUndef() || MI.getNumOperands() > 2) {` . If we are missing the `implicit-def` we have exactly two operands and so we don't add the kill.

Reply:
> However, I'm not sure if the implicit-def is necessary here or not.

Maybe the implicit-def is still necessary at least for the case where the other part of the super register is not allocatable, i.e., once f1 is defined, the vsl1 should also be defined as the phony part can not be allocated.

Reply:
If the phony lane isn't alive at the COPY, I think it's ok we don't have the `implicit-def vsl1`.

Reply:
> If the phony lane isn't alive at the COPY, I think it's ok we don't have the `implicit-def vsl1`.

I think that is what is happening here. I took a look at one of the smaller tests `combine-fneg.ll` to see and the MIR looks like this:
```
  liveins: $f1, $v2, $v3                                                                                                                                                                                        
    undef %3.sub_64:vsrc = COPY $f1                                                                                    
    %4:vsrc = XXPERMDI killed %3:vsrc, %3:vsrc, 0   // This is the only use of %3
```
The only reason we use the VSRC register is because that is the type of register that XXPERMDI requires. The other lane is never used and the value in `$f1` is just splatted into `%4`.